set_transformers/tensorflow/layers.py

```diff
@@ -0,0 +1,456 @@
+        where H = LayerNorm(X + Multihead(X, Y, Y))
+
+        Parameters
+        ----------
+        attention_axes : Tuple[int, ...]
+            determines the axes on which attention is performed
+        multi_head_attention_kwargs : Dict
+            keras layer kwargs
+        rff_kwargs : Dict
+            RFF class kwargs
+        layer_normalization_h_kwargs : Dict
+            keras layers kwargs
+        layer_normalization_out_kwargs : Dict
+            keras layer kwargs
+        normalize : bool, optional
+            if True, use layer_normalization layers as part of
+            the architecture (as per the original paper),
+            by default True
+        """
+        super().__init__(**kwargs)
+
+        self.multi_head = MultiHeadAttention(
+            **multi_head_attention_kwargs,
+            attention_axes=attention_axes
+        )
+
+        self.rff = RFF(
+            **rff_kwargs
+        )
+
+        self.normalize = normalize
+        if normalize:
+            self.layer_norm_h = LayerNormalization(
+                **layer_normalization_h_kwargs
+            )
+
+            self.layer_norm_out = LayerNormalization(
+                **layer_normalization_out_kwargs
+            )
+
+    def call(
+        self,
+        x: tf.Tensor,
+        y: tf.Tensor
+    ) -> tf.Tensor:
+        """MAB(X, Y) = LayerNorm(H + rFF(H))
+        where H = LayerNorm(X + Multihead(X, Y, Y))
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            Query, shape = (batch_size, n, d)
+        y : tf.Tensor
+            Key and Value, shape = (batch_size, n', d)
+
+        Returns
+        -------
+        tf.Tensor
+            shape = (batch_size, n, d)
+        """
+        h = x + self.multi_head(x, y, y)
+        if self.normalize:
+            h = self.layer_norm_h(h)
+
+        out = h + self.rff(h)
+        if self.normalize:
+            out = self.layer_norm_out(out)
+
+        return out
+
+
+class SAB(MAB):
+
+    def call(
+        self,
+        x: tf.Tensor
+    ) -> tf.Tensor:
+        """SAB(X) = MAB(X, X)
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            Query, Key, Value, shape = (batch_size, n, d)
+
+        Returns
+        -------
+        tf.Tensor
+            shape = (batch_size, n, d)
+        """
+        return super().call(x, x)
+
+
+class ISAB(tf.keras.layers.Layer):
+
+    def __init__(
+        self,
+        attention_axes: Tuple[int, ...],
+        m: int,
+        d: int,
+        mab_h_kwargs: Dict,
+        mab_out_kwargs: Dict,
+        **kwargs
+    ):
+        """Induced Self Attention Block
+        ISAB (X) = MAB(X, H) ∈ R n×d
+        where H = MAB(I, X) ∈ R m×d
+
+        Parameters
+        ----------
+        attention_axes : Tuple[int, ...]
+            determines the axes on which attention is performed
+        m : int
+            number of inducing points
+        d : int
+            data point dimension
```

Comment:
Same remark as previously: from an external point of view, `m` seems unnatural for the number of inducing points (`I` or `H` or `n` in paper)